Create interface for accessing the training code (#12)
instructlab/training: README.md

otherwise require distributed training configuration, you can
just create a default object:

```python
run_training(
	torchrun_args=TorchrunArgs(),
	training_args=TrainingArgs(
		# ...
	),
)
```

However, if you want to specify a more complex configuration,
we currently expose all of the options that [torchrun accepts
today](https://pytorch.org/docs/stable/elastic/run.html#definitions). 

> ![NOTE]
> For more information about the `torchrun` arguments, please consult the [torchrun documentation](https://pytorch.org/docs/stable/elastic/run.html#definitions).

For example, in a 8-GPU, 2-machine system, we would 
specify the following torchrun config:

```python
# on machine 1
torchrun_args = TorchrunArgs(
	nnodes = 2, # number of machines 
	nproc_per_node = 4, # num GPUs per machine
	node_rank = 0, # node rank for this machine
	rdzv_id = 123,
	rdzv_endpoint = '127.0.0.1:12345'
)

run_training(
	torchrun_args=torchrun_args,
	training_args=training_args
)
```

```python
# on machine 2
torchrun_args = TorchrunArgs(
    nnodes = 2, # number of machines 
    nproc_per_node = 4, # num GPUs per machine
    node_rank = 1, # node rank for this machine
    rdzv_id = 123,

>     rdzv_endpoint = '127.0.0.1:12345'

Reviewer: very minor nit- the second machine wouldn't refer to localhost, it'd refer to the public IP of the main machine.

Author: Good catch!

Author: Updated

# InstructLab Training Library

In order to simplify the process of fine-tuning models through the LAB 
method, this library provides a simple training interface.

## Installation

To get started with the library, you must clone this repo and install it from source via `pip`:

```bash
# clone the repo and switch to the directory
git clone https://github.com/instructlab/training
cd training

# install the library
pip install .
```

For development, install it instead with `pip install -e .` instead
to make local changes while using this library elsewhere.

### Installing Additional NVIDIA packages

We make use of `flash-attn` and other packages which rely on NVIDIA-specific
CUDA tooling to be installed.

If you are using NVIDIA hardware with CUDA, please install the additional dependencies via:

```bash
# for a regular install
pip install .[cuda]

# or, for an editable install (development)
pip install -e .[cuda]
```

## Usage

> Using the library is fairly straightforward, simply import the necessary items,
> namely:

Reviewer 2: Minor suggestion:

```suggestion
To use the library, import the necessary items:
```

Author: I love it

```py
from instructlab.training import (
	run_training,
	TorchrunArgs,
	TrainingArgs,
	DeepSpeedOptions
)
```

Then, define the training arguments which will serve as the
parameters for our training run:

```py
# define training-specific arguments
training_args = TrainingArgs(
	# define data-specific arguments
	model_path = "ibm-granite/granite-7b-base",
	data_path = "path/to/dataset.jsonl",
	ckpt_output_dir = "data/saved_checkpoints",
	data_output_dir = "data/outputs",

	# define model-trianing parameters
	max_seq_len = 4096,
	max_batch_len = 60000,
	num_epochs = 10,
	effective_batch_size = 3840,
	save_samples = 250000,
	learning_rate = 2e-6,
	warmup_steps = 800,
	is_padding_free = True, # set this to true when using Granite-based models
	random_seed = 42,
)
```

We'll also need to define the settings for running a multi-process job

> via `torchrun`. To do this, simply create a `TorchrunArgs` object.

Reviewer 2: Minor suggestion:

```suggestion
via `torchrun`. To do this, create a `TorchrunArgs` object.
```

Author: 👨‍🍳 🤌
